cpp/src/io/json/json_gpu.cu

```diff
@@ -418,84 +415,151 @@ __device__ __inline__ bool is_like_float(
                                                cudf::size_type num_records,
                                                const data_type *dtypes,
                                                ParseOptions opts,
+                                               col_map_type *col_map,
                                                void *const *output_columns,
-                                               int num_columns,
+                                               cudf::size_type num_columns,
                                                bitmask_type *const *valid_fields,
                                                cudf::size_type *num_valid_fields)
 {
-  const long rec_id = threadIdx.x + (blockDim.x * blockIdx.x);
+  const auto rec_id = threadIdx.x + (blockDim.x * blockIdx.x);
   if (rec_id >= num_records) return;
 
-  long start = rec_starts[rec_id];
-  // has the same semantics as end() in STL containers (one past last element)
-  long stop = ((rec_id < num_records - 1) ? rec_starts[rec_id + 1] : data_size);
+  auto const row_data_range = get_row_data_range(data, data_size, rec_starts, num_records, rec_id);
 
-  limit_range_to_brackets(data, start, stop);
-  const bool is_object = (data[start - 1] == '{');
+  auto current = row_data_range.first;
+  for (size_type input_field_index = 0;
+       input_field_index < num_columns && current < row_data_range.second;
+       input_field_index++) {
+    auto const desc =
+      next_field_descriptor(current, row_data_range.second, opts, input_field_index, col_map);
+    auto const value_len = desc.value_end - desc.value_begin;
+
+    current = desc.value_end + 1;
 
-  for (int col = 0; col < num_columns && start < stop; col++) {
-    if (is_object) { start = seek_field_name_end(data, opts, start, stop); }
-    // field_end is at the next delimiter/newline
-    const long field_end = cudf::io::gpu::seek_field_end(data, opts, start, stop);
-    long field_data_last = field_end - 1;
-    // Modify start & end to ignore whitespace and quotechars
-    trim_field_start_end(data, &start, &field_data_last, opts.quotechar);
     // Empty fields are not legal values
-    if (start <= field_data_last &&
-        !serializedTrieContains(opts.naValuesTrie, data + start, field_end - start)) {
+    if (value_len > 0 && !serializedTrieContains(opts.naValuesTrie, desc.value_begin, value_len)) {
       // Type dispatcher does not handle strings
-      if (dtypes[col].id() == type_id::STRING) {
-        auto str_list           = static_cast<string_pair *>(output_columns[col]);
-        str_list[rec_id].first  = data + start;
-        str_list[rec_id].second = field_data_last - start + 1;
+      if (dtypes[desc.column].id() == type_id::STRING) {
+        auto str_list           = static_cast<string_pair *>(output_columns[desc.column]);
+        str_list[rec_id].first  = desc.value_begin;
+        str_list[rec_id].second = value_len;
```

Comment:
I think you can use `std::tie` here, i.e.

Reply:
How can I use `std::tie` here? 
`str_list[rec_id] = std::tie(desc.value_begin, value_len); `
does not work because str_list elem is a pair, not a tuple.

Reply:
Hmmm yea i got confused. Not `std::tie`, just regular `{}` for pair construction. Seems to not be allowed in device code though? ¯\_(ツ)_/¯
```cpp
str_list[rec_id] = {desc.value_begin, value_len};
```